src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimizedFn.java

```diff
@@ -0,0 +1,280 @@
+            return new FlatMapFunction<ContextShard, ContextShard>() {
+                private static final long serialVersionUID = 1L;
+                @Override
+                public Iterable<ContextShard> call(ContextShard contextShard) throws Exception {
+                    final SubdivideAndFillReadsIterator iterator = new SubdivideAndFillReadsIterator(bam, auth, outputShardSize, margin, optFilter, contextShard);
+                    return new Iterable<ContextShard>() {
+                        @Override
+                        public Iterator<ContextShard> iterator() {
+                            try {
+                                return new SubdivideAndFillReadsIterator(bam, auth, outputShardSize, margin, optFilter, contextShard);
+                            } catch (Exception x) {
+                                throw new RuntimeException(x);
+                            }
+                        }
+                    };
+                }
+            };
+    }
+
+    /**
+     * Given a shard that has reads and variants, query Google Genomics' Reference server and get reference info
+     * (including an extra margin on either side), and fill that and the correct variants into readContext.
+     */
+    public static ContextShard fillContext(ReferenceMultiSource refSource, ContextShard shard) {
+        if (null==shard) return null;
+        // use the function to make sure we get the exact correct amount of reference bases
+        int start = Integer.MAX_VALUE;
+        int end = Integer.MIN_VALUE;
+        SerializableFunction<GATKRead, SimpleInterval> referenceWindowFunction = refSource.getReferenceWindowFunction();
+        for (GATKRead r : shard.reads) {
+            SimpleInterval readRefs = referenceWindowFunction.apply(r);
+            start = Math.min(readRefs.getStart(), start);
+            end = Math.max(readRefs.getEnd(), end);
+        }
+
+        if (start==Integer.MAX_VALUE) {
+            // there are no reads in this shard, so we're going to remove it
+            return null;
+        }
+
+        SimpleInterval refInterval = new SimpleInterval(shard.interval.getContig(), start, end);
+        ReferenceBases refBases;
+        try {
+            refBases = refSource.getReferenceBases(null, refInterval);
+        } catch (IOException x) {
+            throw new GATKException("Unable to read the reference");
+        }
+
+        ArrayList<ReadContextData> readContext = new ArrayList<>();
+        for (GATKRead r : shard.reads) {
+            SimpleInterval readInterval = new SimpleInterval(r);
+            ArrayList<Variant> variantsOverlappingThisRead = shard.variantsOverlapping(readInterval);
+            // we pass all the bases. That's better because this way it's just a shared
+            // pointer instead of being an array copy. Downstream processing is fine with having
+            // extra bases (it expects a few, actually).
```

Comment:
Actually since it's a shared pointer within the object this also serializes just fine for all reasonable forms of serialization (including Java serialization which is what I was using).